File: midolman/src/main/java/org/midonet/cluster/data/TunnelZone.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.cluster.data;

import java.util.UUID;
import javax.annotation.Nonnull;

import org.codehaus.jackson.annotate.JsonProperty;

import org.midonet.packets.IPv4Addr;

public class TunnelZone extends Entity.Base<UUID, TunnelZone.Data, TunnelZone> {

    public static enum Type {
        gre, vxlan, vtep
    }

    public Type getType() {
        return getData().type;
    }

    public TunnelZone setType(Type type) {
        getData().type = type;
        return self();
    }

    public TunnelZone(UUID uuid, @Nonnull Data data) {
        super(uuid, data);
    }

    public TunnelZone() {
        super(null, new Data());
    }

    @Override
    protected TunnelZone self() {
        return this;
    }

    public TunnelZone setName(String name) {
        getData().name = name;
        return self();
    }

    public String getName() {
        return getData().name;
    }

    public static class Data {
        String name;

        @JsonProperty("@type")
        Type type;

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Data data = (Data) o;

            if (name != null ? !name.equals(data.name) : data.name != null)
                return false;

            if (type != null ? !type.equals(data.type) : data.type != null)
                return false;

            return true;
        }

        @Override
        public int hashCode() {
            int result = name != null ? name.hashCode() : 0;
            return 31 * result +
                   (type != null ? type.hashCode() : 0);
        }

        @Override
        public String toString() {
            return "Data{" +
                "name='" + name + '\'' +
                ",type='" + type + '\'' +
                '}';
        }
    }


    public static class HostConfig extends Entity.Base<UUID, HostConfig.Data, HostConfig> {

        public HostConfig(UUID uuid, @Nonnull HostConfig.Data hostConfigData) {
            super(uuid, hostConfigData);
        }

        public HostConfig(UUID uuid) {
            super(uuid, new Data());
        }

        public HostConfig() {
            super(null, new Data());
        }

        public HostConfig setIp(IPv4Addr ip) {
            getData().ip = ip;
            return self();
        }

        @Override
        public HostConfig self() {
            return this;
        }

        public IPv4Addr getIp() {
            return getData().ip;
        }

        /*
         * For backwards compatibility, since zone members are not really typed
         * anymore, hard-code type to gre.
         */
        public String getType() { return "gre"; }

        public void setType(String type) {
            // IGNORE
        }

        public static class Data {
            IPv4Addr ip;

            @JsonProperty("@type")
            final String type = "gre";

            @Override
            public boolean equals(Object o) {
                if (this == o) return true;
                if (o == null || getClass() != o.getClass()) return false;

                Data data = (Data) o;

                if (ip != null ? !ip.equals(data.ip) : data.ip != null)
                    return false;

                return true;
            }

            @Override
            public int hashCode() {
                return ip != null ? ip.hashCode() : 0;
            }

            @Override
            public String toString() {
                return "Data{" + "ip=" + ip + '}';
            }
        }
    }
}

File: python-midonetclient/src/midonetclient/neutron/tunnel_zone.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright (c) 2014 Midokura Europe SARL, All Rights Reserved.
# All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

import logging

from midonetclient import url_provider
from midonetclient import util
from midonetclient import vendor_media_type as mt

LOG = logging.getLogger(__name__)


class TunnelZoneUrlProviderMixin(url_provider.UrlProviderMixin):
    """Tunnel Zone URL provider mixin

    This mixin provides URLs for tunnel zones.
    """

    def tunnel_zone_url(self, tz_id):
        return self.template_url("tunnelZoneTemplate", tz_id)

    def tunnel_zones_url(self):
        return self.resource_url("tunnelZones")

    def tunnel_zone_host_url(self, tz_id, tz_host_id):
        return self.tunnel_zone_hosts_url(tz_id) + "/" + str(tz_host_id)

    def tunnel_zone_hosts_url(self, tz_id):
        return self.tunnel_zone_url(tz_id) + "/hosts"


class TunnelZoneClientMixin(TunnelZoneUrlProviderMixin):
    """Tunnel Zone mixin

    Mixin that defines all the Neutron tunnel zone operations in MidoNet API.
    """

    @util.convert_case
    def create_tunnel_zone(self, tz):
        LOG.info("create_tunnel_zone %r", tz)
        return self.client.post(self.tunnel_zones_url(),
                                mt.APPLICATION_TUNNEL_ZONE_JSON, body=tz)

    def delete_tunnel_zone(self, tz_id):
        LOG.info("delete_tunnel_zone %r", tz_id)
        self.client.delete(self.tunnel_zone_url(tz_id))

    @util.convert_case
    def get_tunnel_zone(self, tz_id, fields=None):
        LOG.info("get_tunnel_zone %r", tz_id)
        return self.client.get(self.tunnel_zone_url(tz_id),
                               mt.APPLICATION_TUNNEL_ZONE_JSON)

    @util.convert_case
    def get_tunnel_zones(self, filters=None, fields=None, sorts=None,
                         limit=None, marker=None, page_reverse=False):
        LOG.info("get_tunnel_zones")
        return self.client.get(self.tunnel_zones_url(),
                               mt.APPLICATION_TUNNEL_ZONE_COLLECTION_JSON)

    @util.convert_case
    def update_tunnel_zone(self, tz):
        LOG.info("update_tunnel_zone %r", tz)
        return self.client.put(self.tunnel_zone_url(tz["id"]),
                               mt.APPLICATION_TUNNEL_ZONE_JSON, tz)

    @util.convert_case
    def create_tunnel_zone_host(self, tz_host):
        LOG.info("create_tunnel_zone_host %r", tz_host)
        # convert_case converted to camel
        return self.client.post(
            self.tunnel_zone_hosts_url(tz_host["tunnelZoneId"]),
            mt.APPLICATION_TUNNEL_ZONE_HOST_JSON, body=tz_host)

    def delete_tunnel_zone_host(self, tz_id, host_id):
        LOG.info("delete_tunnel_zone_host %r %r", (tz_id, host_id))
        self.client.delete(self.tunnel_zone_host_url(tz_id, host_id))

    @util.convert_case
    def get_tunnel_zone_host(self, tz_id, host_id):
        LOG.info("get_tunnel_zone_host %r %r", (tz_id, host_id))
        return self.client.get(self.tunnel_zone_host_url(tz_id, host_id),
                               mt.APPLICATION_TUNNEL_ZONE_HOST_JSON)

    @util.convert_case
    def get_tunnel_zone_hosts(self, tz_id):
        LOG.info("get_tunnel_zone_hosts %r", tz_id)
        return self.client.get(self.tunnel_zone_hosts_url(tz_id),
                               mt.APPLICATION_TUNNEL_ZONE_HOST_COLLECTION_JSON)
